x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java

```diff
@@ -0,0 +1,158 @@
+        if (ciphertextLength % encryptedPacketLength != 0) {
+            decryptedSize += (ciphertextLength % encryptedPacketLength) - EncryptedRepository.GCM_IV_LENGTH_IN_BYTES
+                    - EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES;
+        }
+        return decryptedSize;
+    }
+
+    public DecryptionPacketsInputStream(InputStream source, SecretKey secretKey, int nonce, int packetLength) {
+        this.source = Objects.requireNonNull(source);
+        this.secretKey = Objects.requireNonNull(secretKey);
+        this.nonce = nonce;
+        if (packetLength <= 0 || packetLength >= EncryptedRepository.MAX_PACKET_LENGTH_IN_BYTES) {
+            throw new IllegalArgumentException("Invalid packet length [" + packetLength + "]");
+        }
+        this.packetLength = packetLength;
+        this.packet = new byte[packetLength + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES];
+        this.iv = new byte[EncryptedRepository.GCM_IV_LENGTH_IN_BYTES];
+        this.hasNext = true;
+        this.counter = EncryptedRepository.PACKET_START_COUNTER;
+    }
+
+    @Override
+    InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+        if (currentComponentIn != null && currentComponentIn.read() != -1) {
+            throw new IllegalStateException("Stream for previous packet has not been fully processed");
+        }
+        if (false == hasNext) {
+            return null;
+        }
+        PrefixInputStream packetInputStream = new PrefixInputStream(source,
+                packetLength + EncryptedRepository.GCM_IV_LENGTH_IN_BYTES + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES,
+                false);
+        int currentPacketLength = decrypt(packetInputStream);
+        // only the last packet is shorter, so this must be the last packet
+        if (currentPacketLength != packetLength) {
+            hasNext = false;
+        }
+        return new ByteArrayInputStream(packet, 0, currentPacketLength);
+    }
+
+    @Override
+    public boolean markSupported() {
+        return false;
+    }
+
+    @Override
+    public void mark(int readlimit) {
+    }
+
+    @Override
+    public void reset() throws IOException {
+        throw new IOException("Mark/reset not supported");
+    }
+
+    private int decrypt(PrefixInputStream packetInputStream) throws IOException {
+        if (packetInputStream.read(iv) != iv.length) {
+            throw new IOException("Error while reading the heading IV of the packet");
+        }
+        ByteBuffer ivBuffer = ByteBuffer.wrap(iv).order(ByteOrder.LITTLE_ENDIAN);
+        if (ivBuffer.getInt(0) != nonce || ivBuffer.getLong(4) != counter++) {
+            throw new IOException("Invalid packet IV");
+        }
+        int packetLength = packetInputStream.read(packet);
+        if (packetLength < EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES) {
+            throw new IOException("Error while reading the packet");
+        }
```

Comment:
Is there a max packetLength check we can do here?

Reply:
Not really, the maximum packet length is implicitly coded in the size of the `packet` array, which does have a check for the maximum packet length on allocation.

Reply:
But I did discover an error here, which I've since fixed:
`int packetLength = packetInputStream.read(packet);` should be `int packetLength = packetInputStream.readNBytes(packet, 0, packet.length);` because the whole complete packet must be read, not only a part of it, and plain `read` does not guarantee the read length.
It was not discovered by tests because the ciphertext is backed by the `ByteArrayInputStream` which always returns the requested byte count for reads. I've changed the success tests to account for input streams that return fewer bytes: 97f591791877f64e37d5b1